In [1]:
# Load packages
from costs_benefits_ssp.cb_calculate import CostBenefits
import pandas as pd
import os
import pathlib
import numpy as np


In [2]:
# Define paths
SSP_PATH = pathlib.Path(os.getcwd()).parent.parent.as_posix()
SSP_RUN  = os.path.join(SSP_PATH, "ssp_run_output/sisepuede_summary_results_run_sisepuede_run_2026-04-30T22;47;01.605912/nwr")


CB_DEFAULT_DEFINITION_PATH = os.path.join(SSP_PATH,"cost-benefits/cb_cost_factors")
CB_DEFAULT_DEFINITION_FILE_PATH = os.path.join(CB_DEFAULT_DEFINITION_PATH, "cb_config_params.xlsx")

CB_OUTPUT = os.path.join(SSP_PATH, "Tableau/data")

In [3]:
CB_OUTPUT

'/Users/alexa/Projects/ssp_uganda_data/ssp_modeling/Tableau/data'

In [4]:
CB_DEFAULT_DEFINITION_PATH

'/Users/alexa/Projects/ssp_uganda_data/ssp_modeling/cost-benefits/cb_cost_factors'

In [5]:
# Load data
ssp_data = pd.read_csv(os.path.join(SSP_RUN, "uganda.csv"))
att_primary = pd.read_csv(os.path.join(SSP_RUN,"ATTRIBUTE_PRIMARY.csv"))
att_strategy = pd.read_csv(os.path.join(SSP_RUN,"ATTRIBUTE_STRATEGY.csv"))
strategy_code_base = "BASE"

In [6]:
att_primary

,primary_id,design_id,strategy_id,future_id
0,0,0,0,0
1,70070,0,6004,0
2,75075,0,6009,0
3,77077,0,6011,0
4,78078,0,6012,0


In [7]:
# Instantiate CostBenefits object
cb = CostBenefits(ssp_data, att_primary, att_strategy, strategy_code_base)

# The export_db_to_excel method saves the initial configuration of the cost tables to an excel file.
# Each sheet represents a table in the cost and benefit program database.
# If the Excel file name is not given, the file will be saved with the default name cb_config_params.xlsx on the current python session.

#cb.export_db_to_excel(CB_DEFAULT_DEFINITION_FILE_PATH)

# Once that the excel file has been updated, we can reload it in order to update the cost factors database
cb.load_cb_parameters(CB_DEFAULT_DEFINITION_FILE_PATH)

The TX TX:FRST:INCREASE_SEQUESTRATION_NZ is missing on AttTransformationCode
The TX TX:LNDU:DEC_WETLAND_LOSS_NZ is missing on AttTransformationCode
The TX TX:LNDU:SET_WETLANDS_MINIMUM_NZ is missing on AttTransformationCode
The TX TX:SCOE:INC_EFFICIENCY_HEAT_NZ is missing on AttTransformationCode
The TX TX:TRNS:SHIFT_FUEL_POWERED_BIKES_STRATEGY_NZ is missing on AttTransformationCode
The TX TX:FRST:INCREASE_SEQUESTRATION_NDC_25 is missing on AttTransformationCode
The TX TX:LNDU:DEC_WETLAND_LOSS_NDC_25 is missing on AttTransformationCode
The TX TX:LNDU:SET_WETLANDS_MINIMUM_NDC_25 is missing on AttTransformationCode
The TX TX:SCOE:INC_EFFICIENCY_HEAT_NDC_25 is missing on AttTransformationCode
The TX TX:TRNS:SHIFT_FUEL_POWERED_BIKES_STRATEGY_NDC_25 is missing on AttTransformationCode
The TX TX:SCOE:INC_EFFICIENCY_HEAT_NDC_25 is missing on AttTransformationCode
The TX TX:LNDU:DEC_WETLAND_LOSS_NDC_25 is missing on AttTransformationCode
The TX TX:LNDU:SET_WETLANDS_MINIMUM_NDC_25 is missing on 

In [8]:
# Compute System Costs
results_system = cb.compute_system_cost_for_all_strategies()


************************************
*Strategy : PFLO:HBLE (0/4)
************************************

---------Costs for: cb:wali:technical_cost:sanitation:unimp_rural.
La variable se evalúa en System Cost
---------Costs for: cb:wali:technical_cost:sanitation:imp_rural.
La variable se evalúa en System Cost
---------Costs for: cb:wali:technical_cost:sanitation:safeman_rural.
La variable se evalúa en System Cost
---------Costs for: cb:wali:technical_cost:sanitation:unimp_urban.
La variable se evalúa en System Cost
---------Costs for: cb:wali:technical_cost:sanitation:imp_urban.
La variable se evalúa en System Cost
---------Costs for: cb:wali:technical_cost:sanitation:safeman_urban.
La variable se evalúa en System Cost
---------Costs for: cb:wali:technical_cost:sanitation:omit_rural.
La variable se evalúa en System Cost
---------Costs for: cb:entc:technical_cost:electricity:capex.
La variable se evalúa en System Cost
---------Costs for: cb:entc:technical_cost:electricity:transmission.
L

In [9]:
# Compute Technical Costs
results_tx = cb.compute_technical_cost_for_all_strategies()


************************************
*Strategy : PFLO:HBLE (0/4)
************************************

---------Costs for: cb:inen:technical_cost:fuel_switch:lo_heat.
La variable se evalúa en Transformation Cost
---------Costs for: cb:inen:technical_savings:fuel_switch:lo_heat.
La variable se evalúa en Transformation Cost
---------Costs for: cb:inen:technical_cost:fuel_switch:hi_heat.
La variable se evalúa en Transformation Cost
---------Costs for: cb:inen:technical_savings:fuel_switch:hi_heat.
La variable se evalúa en Transformation Cost
---------Costs for: cb:inen:technical_cost:fuel_switch:all.
La variable se evalúa en Transformation Cost
---------Costs for: cb:inen:technical_savings:fuel_switch:all.
La variable se evalúa en Transformation Cost
---------Costs for: cb:inen:technical_cost:efficiency:X.
La variable se evalúa en Transformation Cost
---------Costs for: cb:trns:technical_cost:electrification:LDV.
La variable se evalúa en Transformation Cost
---------Costs for: cb:trns:te

/opt/anaconda3/envs/ssp_uganda_env/lib/python3.11/site-packages/costs_benefits_ssp/cb_calculate.py:856: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tmp["difference_variable"] = cb_orm.diff_var
/opt/anaconda3/envs/ssp_uganda_env/lib/python3.11/site-packages/costs_benefits_ssp/cb_calculate.py:857: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tmp["difference_value"] = data_merged["difference"]
/opt/anaconda3/envs/ssp_uganda_env/lib/python3.11/site-packages/costs_benefits_ssp/cb_calculate.py:858: SettingWith

---------Costs for: cb:scoe:technical_cost:efficiency:building_shell.
La variable se evalúa en Transformation Cost
---------Costs for: cb:entc:technical_cost:loss_reduction:electricity.
La variable se evalúa en Transformation Cost
---------Costs for: cb:ippu:technical_cost:clinker:X.
La variable se evalúa en Transformation Cost
---------Costs for: cb:ippu:technical_cost:abating_gases:n2o.
La variable se evalúa en Transformation Cost
---------Costs for: cb:ippu:technical_cost:abating_gases:pfcs.
La variable se evalúa en Transformation Cost
---------Costs for: cb:ippu:technical_cost:abating_gases:other_fgases.
La variable se evalúa en Transformation Cost
---------Costs for: cb:fgtv:technical_cost:flaring:X.
La variable se evalúa en Transformation Cost
---------Costs for: cb:fgtv:technical_cost:leaks:X.
La variable se evalúa en Transformation Cost
---------Costs for: cb:waso:technical_cost:consumer_food_waste:X.
La variable se evalúa en Transformation Cost
---------Costs for: cb:waso:cons

/opt/anaconda3/envs/ssp_uganda_env/lib/python3.11/site-packages/costs_benefits_ssp/cb_calculate.py:1067: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_output['variable_value_baseline'] = 0
/opt/anaconda3/envs/ssp_uganda_env/lib/python3.11/site-packages/costs_benefits_ssp/cb_calculate.py:1070: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_output['variable_value_pathway'] = data_merged["difference_value"]


---------Costs for: cb:entc:technical_savings:electricity:opex.
La variable se evalúa en Transformation Cost

************************************
*Strategy : PFLO:NDC_25_UNCONDITIONAL (2/4)
************************************

---------Costs for: cb:scoe:technical_cost:fuel_switch:heat_pumps.
La variable se evalúa en Transformation Cost
---------Costs for: cb:scoe:technical_savings:fuel_switch:heat_pumps.
La variable se evalúa en Transformation Cost
---------Costs for: cb:scoe:technical_cost:efficiency:building_shell.
La variable se evalúa en Transformation Cost
---------Costs for: cb:waso:technical_cost:biogas_capture.
La variable se evalúa en Transformation Cost

************************************
*Strategy : PFLO:NDC_25_UNCONDITIONAL_CE (3/4)
************************************

---------Costs for: cb:scoe:technical_cost:fuel_switch:heat_pumps.
La variable se evalúa en Transformation Cost
---------Costs for: cb:scoe:technical_savings:fuel_switch:heat_pumps.
La variable se evalú

/opt/anaconda3/envs/ssp_uganda_env/lib/python3.11/site-packages/costs_benefits_ssp/cb_calculate.py:1067: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_output['variable_value_baseline'] = 0
/opt/anaconda3/envs/ssp_uganda_env/lib/python3.11/site-packages/costs_benefits_ssp/cb_calculate.py:1070: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_output['variable_value_pathway'] = data_merged["difference_value"]


In [10]:
# Combine results
results_all = pd.concat([results_system, results_tx], ignore_index = True)

#-------------POST PROCESS SIMULATION RESULTS---------------
# Post process interactions among strategies that affect the same variables
results_all_pp = cb.cb_process_interactions(results_all)

# SHIFT any stray costs incurred from 2015 to 2025 to 2025 and 2035
results_all_pp_shifted = cb.cb_shift_costs(results_all_pp)

Resolving Interactions in SCOE : TX:SCOE:INC_EFFICIENCY_APPLIANCE, TX:SCOE:SHIFT_FUEL_HEAT, TX:SCOE:SHIFT_FUEL_HEAT, TX:SCOE:DEC_DEMAND_HEAT 
Resolving Interactions in INEN : TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:INC_EFFICIENCY_ENERGY 
Resolving Interactions in SCOE : TX:SCOE:SHIFT_FUEL_HEAT, TX:SCOE:SHIFT_FUEL_HEAT, TX:SCOE:DEC_DEMAND_HEAT 
Resolving Interactions in INEN : TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:INC_EFFICIENCY_ENERGY 
Resolving Interactions in SCOE : TX:SCOE:SHIFT_FUEL_HEAT, TX:SCOE:SHIFT_FUEL_HEAT, TX:SCOE:DEC_DEMAND_HEAT 
Resolving Interactions in INEN :  
No interactions, skipping... PFLO:NDC_25_UNCONDITIONAL
Resolving Interactions in SCOE : TX:SCOE:SHIFT_FUEL_HEAT, TX:SCOE:SHIFT_FUEL_HEAT, TX:SCOE:DEC_DEMAND_HEAT 
Resolving Intera

/opt/anaconda3/envs/ssp_uganda_env/lib/python3.11/site-packages/costs_benefits_ssp/cb_calculate.py:660: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  res_pre2025["variable"] = res_pre2025["variable"] + "_shifted" + (res_pre2025["time_period"]+SSP_GLOBAL_TIME_PERIOD_0).astype(str)#create a new variable so they can be recognized as shifted costs
/opt/anaconda3/envs/ssp_uganda_env/lib/python3.11/site-packages/costs_benefits_ssp/cb_calculate.py:661: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  res_pre2025["tim

In [11]:
results_all_pp_shifted.head()

,strategy_code,future_id,region,time_period,difference_variable,variable_value_baseline,variable_value_pathway,difference_value,variable,value
0,PFLO:HBLE,0,uganda,4,pop_unimproved_rural,1.932346e+07,1.932346e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0
1,PFLO:HBLE,0,uganda,5,pop_unimproved_rural,1.951974e+07,1.951974e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0
2,PFLO:HBLE,0,uganda,6,pop_unimproved_rural,1.896047e+07,1.896047e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0
3,PFLO:HBLE,0,uganda,7,pop_unimproved_rural,1.837001e+07,1.837001e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0
4,PFLO:HBLE,0,uganda,8,pop_unimproved_rural,1.774752e+07,1.774752e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0


In [12]:
ssp_data.head()

,primary_id,region,time_period,area_agrc_crops_bevs_and_spices,area_agrc_crops_cereals,area_agrc_crops_fibers,area_agrc_crops_fruits,area_agrc_crops_herbs_and_other_perennial_crops,area_agrc_crops_nuts,area_agrc_crops_other_annual,...,yf_agrc_herbs_and_other_perennial_crops_tonne_ha,yf_agrc_nuts_tonne_ha,yf_agrc_other_annual_tonne_ha,yf_agrc_other_woody_perennial_tonne_ha,yf_agrc_pulses_tonne_ha,yf_agrc_rice_tonne_ha,yf_agrc_sugar_cane_tonne_ha,yf_agrc_tubers_tonne_ha,yf_agrc_vegetables_and_vines_tonne_ha,yf_lndu_supremum_pastures_tonne_per_ha
0,0,uganda,4,634960.593061,2.025945e+06,156073.139455,1.131815e+06,505.429083,513832.892470,2.915383e+06,...,2.172065,0.458120,1.528325,0.3674,0.799814,2.970174,77.726965,4.805888,4.035314,92.81
1,0,uganda,5,640126.452982,2.042427e+06,157342.906406,1.141024e+06,509.541112,518013.291655,2.939102e+06,...,2.196794,0.611231,1.681739,0.3674,0.805568,3.032345,78.023706,4.440105,4.013004,92.81
2,0,uganda,6,642767.262368,2.050853e+06,157992.016628,1.145731e+06,511.643198,520150.329354,2.951227e+06,...,2.210866,0.841393,2.293907,0.3674,0.807398,3.080626,61.430414,4.911305,4.023409,92.81
3,0,uganda,7,645322.479649,2.059006e+06,158620.088334,1.150285e+06,513.677153,522218.102852,2.962959e+06,...,2.225270,0.400800,2.699337,0.3674,0.810044,3.094956,77.255948,4.924276,4.030111,92.81
4,0,uganda,8,647798.760608,2.066907e+06,159228.757514,1.154699e+06,515.648274,524221.998246,2.974329e+06,...,2.225270,0.400800,2.699337,0.3674,0.810044,3.094956,77.255948,4.924276,4.030111,92.81


# Indoor air pollution benef

Health co-benefit from reduced household air pollution

To monetize the health benefits of reducing indoor air pollution from the use of traditional biomass stoves, we drew on estimates from the BAR-HAP model (The Benefits of Action to Reduce Household Air Pollution; WHO, 2021).
The study reports cost-effectiveness ratios ranging from 674 – 869 USD per DALY avoided for clean cooking interventions in Nepal, meaning that each avoided disability-adjusted life year (DALY) costs approximately 700 USD.

To adapt this metric to Uganda, we scaled the cost-effectiveness ratio by the relative difference in GDP per capita between Nepal (~1,450 USD) and Uganda (~1,070 USD), yielding an adjusted range of ≈ 500–640 USD per DALY avoided.
Assuming that the transition from solid fuels to clean cooking solutions avoids about 0.1 DALY per person per year, the implied annual health benefit equals 50–64 USD per person.

Given an average household size of five, this corresponds to an estimated 250–320 USD per household per year in health-related economic benefits.

This adjusted value was used in the cost–benefit analysis as the annual household health co-benefit of eliminating exposure to indoor air pollution from firewood and other solid fuels in Uganda.

In [13]:
# --- Merge strategy attributes into ssp_data ---
# Needed to group transitions by strategy and retrieve strategy_code / future_id.
df_ssp = ssp_data.merge(
    att_primary[['primary_id', 'strategy_id', 'future_id']],
    on='primary_id', how='left'
)
df_ssp = df_ssp.merge(
    att_strategy[['strategy_id', 'strategy_code']],
    on='strategy_id', how='left'
)

# Variable maps: SISEPUEDE fuel-fraction columns -> readable transition labels.
# Used to detect which population share leaves (out_of) or enters (into) each fuel.
var_map_out = {
    'frac_scoe_heat_energy_residential_solid_biomass':             'est_out_of_biomass',
    'frac_scoe_heat_energy_residential_electricity':               'est_out_of_elec',
    'frac_scoe_heat_energy_residential_hydrocarbon_gas_liquids':   'est_out_of_gas',
}

var_map_into = {
    'frac_scoe_heat_energy_residential_solid_biomass':             'est_into_biomass',
    'frac_scoe_heat_energy_residential_electricity':               'est_into_elec',
    'frac_scoe_heat_energy_residential_hydrocarbon_gas_liquids':   'est_into_gas',
}

In [14]:
# --- Helper functions to compute period-over-period fuel-fraction transitions ---

def add_out_of(df, time_col, group_cols, var_map):
    """
    out_of(t) = max(frac(t) - frac(t+1), 0)
    Captures the share of population that leaves a given fuel between t and t+1.
    Clipped at 0 so that increases in a fuel do not produce negative 'out_of' values.
    """
    df = df.sort_values(group_cols + [time_col]).copy()
    for frac_var, out_var in var_map.items():
        next_frac = df.groupby(group_cols, sort=False)[frac_var].shift(-1)
        df[out_var] = (df[frac_var] - next_frac).clip(lower=0).fillna(0)
    return df


def add_into(df, time_col, group_cols, var_map):
    """
    into(t) = max(frac(t+1) - frac(t), 0)
    Captures the share of population that enters a given fuel between t and t+1.
    Clipped at 0 so that decreases in a fuel do not produce negative 'into' values.
    """
    df = df.sort_values(group_cols + [time_col]).copy()
    for frac_var, into_var in var_map.items():
        next_frac = df.groupby(group_cols, sort=False)[frac_var].shift(-1)
        df[into_var] = (next_frac - df[frac_var]).clip(lower=0).fillna(0)
    return df

In [15]:
def add_transitions(df):
    """
    Derives mutually exclusive household fuel-switching transitions per period.
    Priority order prevents double-counting:
      1. biomass -> gas  (matched first)
      2. biomass -> elec (remaining biomass leavers)
      3. gas    -> elec  (remaining electricity entrants)
    Each transition value is a fraction of total population.
    """
    df = df.copy()

    # 1. Biomass -> Gas
    df['biomass_to_gas'] = df[['est_out_of_biomass', 'est_into_gas']].min(axis=1)
    rem_out_biomass      = df['est_out_of_biomass'] - df['biomass_to_gas']
    rem_into_gas         = df['est_into_gas']       - df['biomass_to_gas']

    # 2. Biomass -> Elec (with remaining biomass leavers)
    df['biomass_to_elec'] = pd.concat([rem_out_biomass, df['est_into_elec']], axis=1).min(axis=1)
    rem_into_elec         = df['est_into_elec'] - df['biomass_to_elec']

    # 3. Gas -> Elec (with remaining electricity entrants)
    df['gas_to_elec'] = pd.concat([df['est_out_of_gas'], rem_into_elec], axis=1).min(axis=1)

    return df


# Apply: compute out_of, into, and final transitions
df_ssp = add_out_of(df_ssp,  time_col='time_period', group_cols=['strategy_id'], var_map=var_map_out)
df_ssp = add_into(df_ssp,    time_col='time_period', group_cols=['strategy_id', 'region'], var_map=var_map_into)
df_ssp = add_transitions(df_ssp)

df_ssp[['time_period', 'strategy_id', 'biomass_to_gas', 'biomass_to_elec', 'gas_to_elec']].head(10)

,time_period,strategy_id,biomass_to_gas,biomass_to_elec,gas_to_elec
0,4,0,0.009584,0.007447,0.0
1,5,0,0.010887,0.008697,0.0
2,6,0,0.012157,0.010050,0.0
3,7,0,0.013071,0.011286,0.0
4,8,0,0.013467,0.012325,0.0
5,9,0,0.013293,0.013175,0.0
6,10,0,0.012566,0.013891,0.0
7,11,0,0.011346,0.035974,0.0
8,12,0,0.009722,0.038762,0.0
9,13,0,0.007804,0.041568,0.0


In [16]:
# --- Per-capita cost and benefit factors by transition type ---
# Source: WHO BAR-HAP model (2021), scaled to Uganda GDP per capita.
# Four policy instruments are considered (stove subsidy, fuel subsidy, financing,
# behavioral change campaign). Because no data is available on the share of
# households reached by each instrument, factors are averaged across policies
# (unweighted). Values are in PPP 2019 USD per person.
# Costs are negative (government expenditure); benefits are positive (health gains).

df_factors_raw = pd.DataFrame([
    # --- biomass -> gas ---
    #{'transition': 'biomass_to_gas',  'policy': 'subsidy_for_stove',                    'cost_per_capita_ppp_2019': -1.349440118, 'benefit_per_capita_ppp_2019':  4.767680737},
    #{'transition': 'biomass_to_gas',  'policy': 'fuel_subsidy',                         'cost_per_capita_ppp_2019': -3.431367796, 'benefit_per_capita_ppp_2019':  5.341173225},
    {'transition': 'biomass_to_gas',  'policy': 'financing',                             'cost_per_capita_ppp_2019': -6, 'benefit_per_capita_ppp_2019': 37},
    #{'transition': 'biomass_to_gas',  'policy': 'intensive_behavioral_change_campaign',  'cost_per_capita_ppp_2019': -2.012617193, 'benefit_per_capita_ppp_2019':  6.868823172},
    # --- biomass -> elec ---
    #{'transition': 'biomass_to_elec', 'policy': 'subsidy_for_stove',                    'cost_per_capita_ppp_2019': -2.659697106, 'benefit_per_capita_ppp_2019':  9.566241834},
    #{'transition': 'biomass_to_elec', 'policy': 'fuel_subsidy',                         'cost_per_capita_ppp_2019': -6.229455845, 'benefit_per_capita_ppp_2019': 11.330111010},
    {'transition': 'biomass_to_elec', 'policy': 'financing',                             'cost_per_capita_ppp_2019': -19, 'benefit_per_capita_ppp_2019': 57},
    #{'transition': 'biomass_to_elec', 'policy': 'intensive_behavioral_change_campaign',  'cost_per_capita_ppp_2019': -3.535655711, 'benefit_per_capita_ppp_2019': 11.564970400},
    
    {'transition': 'gas_to_elec', 'policy': 'fuel_subsidy',                             'cost_per_capita_ppp_2019': -4, 'benefit_per_capita_ppp_2019': 539},
])

# Average across policies — one factor row per transition type
df_factors = (
    df_factors_raw
    .groupby('transition', as_index=False)
    .agg(
        cost_per_capita_ppp_2019=('cost_per_capita_ppp_2019',    'sum'),
        benefit_per_capita_ppp_2019=('benefit_per_capita_ppp_2019', 'sum'),
    )
)
df_factors

,transition,cost_per_capita_ppp_2019,benefit_per_capita_ppp_2019
0,biomass_to_elec,-19,57
1,biomass_to_gas,-6,37
2,gas_to_elec,-4,539


In [17]:
# --- Monetization ---
# Only biomass_to_gas and biomass_to_elec generate indoor air pollution benefits
# (gas_to_elec does not alter indoor air quality in the same way).

transitions = ['biomass_to_gas', 'biomass_to_elec', 'gas_to_elec']

# Melt to long format: one row per (time_period, strategy_id, transition)
ssp_long = (
    df_ssp
    .melt(
        id_vars=['time_period', 'strategy_id', 'population_gnrl_total'],
        value_vars=transitions,
        var_name='transition',
        value_name='transition_frac',
    )
)

# Population affected = fraction transitioning × total population
ssp_long['population_affected'] = ssp_long['transition_frac'] * ssp_long['population_gnrl_total']

# Merge averaged per-capita factors and compute monetary values (USD PPP 2019)
cb_ap = ssp_long.merge(df_factors, on='transition', how='left')
cb_ap['total_cost']    = cb_ap['population_affected'] * cb_ap['cost_per_capita_ppp_2019']
cb_ap['total_benefit'] = cb_ap['population_affected'] * cb_ap['benefit_per_capita_ppp_2019']
cb_ap['net_benefit']   = cb_ap['total_benefit'] + cb_ap['total_cost']   # cost is negative

cb_ap.head()

,time_period,strategy_id,population_gnrl_total,transition,transition_frac,population_affected,cost_per_capita_ppp_2019,benefit_per_capita_ppp_2019,total_cost,total_benefit,net_benefit
0,4,0,39853824,biomass_to_gas,0.009584,381955.329412,-6,37,-2.291732e+06,1.413235e+07,1.184062e+07
1,5,0,40996690,biomass_to_gas,0.010887,446317.967983,-6,37,-2.677908e+06,1.651376e+07,1.383586e+07
2,6,0,42172327,biomass_to_gas,0.012157,512693.705314,-6,37,-3.076162e+06,1.896967e+07,1.589350e+07
3,7,0,43381679,biomass_to_gas,0.013071,567036.280774,-6,37,-3.402218e+06,2.098034e+07,1.757812e+07
4,8,0,44625710,biomass_to_gas,0.013467,600968.376655,-6,37,-3.605810e+06,2.223583e+07,1.863002e+07


In [18]:
cb_ap[(cb_ap['strategy_id'] == 6004) & (cb_ap['time_period'] == 4)]

,time_period,strategy_id,population_gnrl_total,transition,transition_frac,population_affected,cost_per_capita_ppp_2019,benefit_per_capita_ppp_2019,total_cost,total_benefit,net_benefit
52,4,6004,39853824,biomass_to_gas,0.009584,381955.329412,-6,37,-2.291732e+06,1.413235e+07,1.184062e+07
312,4,6004,39853824,biomass_to_elec,0.007447,296792.096343,-19,57,-5.639050e+06,1.691715e+07,1.127810e+07
572,4,6004,39853824,gas_to_elec,0.000000,0.000000,-4,539,-0.000000e+00,0.000000e+00,0.000000e+00


In [19]:
# --- Aggregate net benefit by (strategy_id, time_period) ---
# Sum net_benefit across both transitions to get a single CBA value per
# strategy/period, then attach strategy metadata for downstream use.

merged = (
    cb_ap
    .groupby(['strategy_id', 'time_period'], as_index=False)['net_benefit']
    .sum()
    .rename(columns={'net_benefit': 'value'})
)

# Exclude baseline (strategy_id == 0); benefits are relative to BAU by construction
merged = merged[merged['strategy_id'] != 0].copy()

# Re-attach strategy_code and future_id from attribute tables
merged = merged.merge(att_primary[['primary_id', 'strategy_id', 'future_id']].drop_duplicates('strategy_id'),
                      on='strategy_id', how='left')
merged = merged.merge(att_strategy[['strategy_id', 'strategy_code']], on='strategy_id', how='left')

In [20]:
merged

,strategy_id,time_period,value,primary_id,future_id,strategy_code
0,6004,4,2.311871e+07,70070,0,PFLO:HBLE
1,6004,5,2.738452e+07,70070,0,PFLO:HBLE
2,6004,6,3.199844e+07,70070,0,PFLO:HBLE
3,6004,7,3.618300e+07,70070,0,PFLO:HBLE
4,6004,8,3.953099e+07,70070,0,PFLO:HBLE
...,...,...,...,...,...,...
203,6012,51,6.791939e+07,78078,0,PFLO:NDC_25_UNCONDITIONAL_CE
204,6012,52,6.663452e+07,78078,0,PFLO:NDC_25_UNCONDITIONAL_CE
205,6012,53,6.492085e+07,78078,0,PFLO:NDC_25_UNCONDITIONAL_CE
206,6012,54,6.285199e+07,78078,0,PFLO:NDC_25_UNCONDITIONAL_CE


In [21]:
# Add metadata columns required by results_all_pp_shifted schema
merged['region']              = 'uganda'
merged['difference_variable'] = 'frac_scoe_heat_energy_residential_solid_biomass'
merged['variable']            = 'cb:scoe:indoor_air_pollution'

# Select only the columns that results_all_pp_shifted expects
merged = merged[['strategy_code', 'future_id', 'region', 'time_period',
                  'difference_variable', 'value', 'variable']]

merged.head()

,strategy_code,future_id,region,time_period,difference_variable,value,variable
0,PFLO:HBLE,0,uganda,4,frac_scoe_heat_energy_residential_solid_biomass,2.311871e+07,cb:scoe:indoor_air_pollution
1,PFLO:HBLE,0,uganda,5,frac_scoe_heat_energy_residential_solid_biomass,2.738452e+07,cb:scoe:indoor_air_pollution
2,PFLO:HBLE,0,uganda,6,frac_scoe_heat_energy_residential_solid_biomass,3.199844e+07,cb:scoe:indoor_air_pollution
3,PFLO:HBLE,0,uganda,7,frac_scoe_heat_energy_residential_solid_biomass,3.618300e+07,cb:scoe:indoor_air_pollution
4,PFLO:HBLE,0,uganda,8,frac_scoe_heat_energy_residential_solid_biomass,3.953099e+07,cb:scoe:indoor_air_pollution


In [22]:
# Optional: save air pollution CBA to CSV for inspection
#merged.to_csv(os.path.join(CB_OUTPUT, 'cba_indoor_air_pollution.csv'), index=False)

In [23]:
# (placeholder — cell reserved for future use)

In [24]:
# (placeholder — cell reserved for future use)

In [25]:
results_all_pp_shifted= pd.concat([results_all_pp_shifted, merged], ignore_index = True)

In [26]:
results_all_pp_shifted.head()

,strategy_code,future_id,region,time_period,difference_variable,variable_value_baseline,variable_value_pathway,difference_value,variable,value
0,PFLO:HBLE,0,uganda,4,pop_unimproved_rural,1.932346e+07,1.932346e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0
1,PFLO:HBLE,0,uganda,5,pop_unimproved_rural,1.951974e+07,1.951974e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0
2,PFLO:HBLE,0,uganda,6,pop_unimproved_rural,1.896047e+07,1.896047e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0
3,PFLO:HBLE,0,uganda,7,pop_unimproved_rural,1.837001e+07,1.837001e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0
4,PFLO:HBLE,0,uganda,8,pop_unimproved_rural,1.774752e+07,1.774752e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0


In [27]:
results_all_pp_shifted.tail()

,strategy_code,future_id,region,time_period,difference_variable,variable_value_baseline,variable_value_pathway,difference_value,variable,value
50435,PFLO:NDC_25_UNCONDITIONAL_CE,0,uganda,51,frac_scoe_heat_energy_residential_solid_biomass,NaN,NaN,NaN,cb:scoe:indoor_air_pollution,6.791939e+07
50436,PFLO:NDC_25_UNCONDITIONAL_CE,0,uganda,52,frac_scoe_heat_energy_residential_solid_biomass,NaN,NaN,NaN,cb:scoe:indoor_air_pollution,6.663452e+07
50437,PFLO:NDC_25_UNCONDITIONAL_CE,0,uganda,53,frac_scoe_heat_energy_residential_solid_biomass,NaN,NaN,NaN,cb:scoe:indoor_air_pollution,6.492085e+07
50438,PFLO:NDC_25_UNCONDITIONAL_CE,0,uganda,54,frac_scoe_heat_energy_residential_solid_biomass,NaN,NaN,NaN,cb:scoe:indoor_air_pollution,6.285199e+07
50439,PFLO:NDC_25_UNCONDITIONAL_CE,0,uganda,55,frac_scoe_heat_energy_residential_solid_biomass,NaN,NaN,NaN,cb:scoe:indoor_air_pollution,0.000000e+00


## Reshape to load into tableau

In [28]:
# --- Read cost-benefit data ---
cb_data = results_all_pp_shifted.copy()

# Split 'variable' into components: name, sector, cb_type, item_1, item_2
# (Assumes exactly 5 colon-separated parts; if there are more colons inside the last field,
# they will be kept in item_2 thanks to n=4)
cb_chars = cb_data["variable"].astype(str).str.split(":", n=4, expand=True)
cb_chars.columns = ["name", "sector", "cb_type", "item_1", "item_2"]
cb_data = pd.concat([cb_data, cb_chars], axis=1)

In [29]:
cb_data.head()

,strategy_code,future_id,region,time_period,difference_variable,variable_value_baseline,variable_value_pathway,difference_value,variable,value,name,sector,cb_type,item_1,item_2
0,PFLO:HBLE,0,uganda,4,pop_unimproved_rural,1.932346e+07,1.932346e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0,cb,wali,technical_cost,sanitation,unimp_rural
1,PFLO:HBLE,0,uganda,5,pop_unimproved_rural,1.951974e+07,1.951974e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0,cb,wali,technical_cost,sanitation,unimp_rural
2,PFLO:HBLE,0,uganda,6,pop_unimproved_rural,1.896047e+07,1.896047e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0,cb,wali,technical_cost,sanitation,unimp_rural
3,PFLO:HBLE,0,uganda,7,pop_unimproved_rural,1.837001e+07,1.837001e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0,cb,wali,technical_cost,sanitation,unimp_rural
4,PFLO:HBLE,0,uganda,8,pop_unimproved_rural,1.774752e+07,1.774752e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0,cb,wali,technical_cost,sanitation,unimp_rural


In [30]:
cb_data['strategy_code'].unique()

array(['PFLO:HBLE', 'PFLO:NDC_25', 'PFLO:NDC_25_UNCONDITIONAL',
       'PFLO:NDC_25_UNCONDITIONAL_CE'], dtype=object)

In [31]:
# Scale value from USD to billions (divide by 1e9)
if "value" in cb_data.columns:
    cb_data["value"] = cb_data["value"] / 1e9

# --- Remove "shifted" entries ---
# Remove rows where item_2 contains "shifted"
cb_data = cb_data[~cb_data["item_2"].astype(str).str.contains("shifted", na=False)]

# Remove any remaining rows where variable contains "shifted2"
cb_data = cb_data[~cb_data["variable"].astype(str).str.contains("shifted2", na=False)]

# --- Add Year column (Year = time_period + 2015) ---
cb_data["Year"] = cb_data["time_period"] + 2015

**This Section Needs Manual Tunnning**

In [32]:
cb_data.strategy_code.unique()

array(['PFLO:HBLE', 'PFLO:NDC_25', 'PFLO:NDC_25_UNCONDITIONAL',
       'PFLO:NDC_25_UNCONDITIONAL_CE'], dtype=object)

In [33]:
# Harmonize strategy name changes with a single mapping (no duplicate replacements)
strategy_name_map = {
    "PFLO:HBLE": "PFLO:HBLE",
    "PFLO:NDC_25_SCOE_ELEC": "PFLO:NDC_25_SCOE_ELEC",
    "PFLO:NDC_25_UNCONDITIONAL": "NDC_25_UNCONDITIONAL",
    "PFLO:NDC_25_UNCONDITIONAL_CE": "PFLO:NDC_25_UNCONDITIONAL_CE"
}

cb_data["strategy"] = (
    cb_data["strategy_code"].astype(str).map(strategy_name_map).fillna(cb_data["strategy_code"])
)
print(cb_data["strategy"].unique())


['PFLO:HBLE' 'PFLO:NDC_25' 'NDC_25_UNCONDITIONAL'
 'PFLO:NDC_25_UNCONDITIONAL_CE']


In [34]:
# --- Strategy ID mapping based on strategy_code ---
def map_strategy_id(code: str):
    if code == "BASE":
        return 0
    if code == "PFLO:HBLE":
        return 6004
    if code == "PFLO:NDC_25":
        return 6009
    # if code == "PFLO:NDC_25_SCOE_ELEC":
    #     return 6010
    if code == "PFLO:NDC_25_UNCONDITIONAL":
        return 6011
    if code == "PFLO:NDC_25_UNCONDITIONAL_CE":
        return 6012
    return code  # keep original if not matched

cb_data["strategy_id"] = cb_data["strategy_code"].astype(str).map(map_strategy_id)

In [35]:
ssp_data.primary_id.unique()

array([    0, 70070, 75075, 77077, 78078])

In [36]:
def map_primary_id(code: str):
    if code == "BASE":
        return 0
    if code == "PFLO:HBLE":
        return 70070
    if code == "PFLO:NDC_25":
        return 75075
    # if code == "PFLO:NDC_25_SCOE_ELEC":
    #     return 760076
    if code == "PFLO:NDC_25_UNCONDITIONAL":
        return 77077
    if code == "PFLO:NDC_25_UNCONDITIONAL_CE":
        return 78078
    return code  # keep original if not matched

cb_data["primary_id"] = cb_data["strategy_code"].astype(str).map(map_primary_id)


**Manual Tunning Ends Here**

In [37]:
# --- Create ids (variable:strategy_id) ---
cb_data["ids"] = cb_data["variable"].astype(str) + ":" + cb_data["strategy_id"].astype(str)

In [38]:
cb_data.head()

,strategy_code,future_id,region,time_period,difference_variable,variable_value_baseline,variable_value_pathway,difference_value,variable,value,name,sector,cb_type,item_1,item_2,Year,strategy,strategy_id,primary_id,ids
0,PFLO:HBLE,0,uganda,4,pop_unimproved_rural,1.932346e+07,1.932346e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0,cb,wali,technical_cost,sanitation,unimp_rural,2019,PFLO:HBLE,6004,70070,cb:wali:technical_cost:sanitation:unimp_rural:...
1,PFLO:HBLE,0,uganda,5,pop_unimproved_rural,1.951974e+07,1.951974e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0,cb,wali,technical_cost,sanitation,unimp_rural,2020,PFLO:HBLE,6004,70070,cb:wali:technical_cost:sanitation:unimp_rural:...
2,PFLO:HBLE,0,uganda,6,pop_unimproved_rural,1.896047e+07,1.896047e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0,cb,wali,technical_cost,sanitation,unimp_rural,2021,PFLO:HBLE,6004,70070,cb:wali:technical_cost:sanitation:unimp_rural:...
3,PFLO:HBLE,0,uganda,7,pop_unimproved_rural,1.837001e+07,1.837001e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0,cb,wali,technical_cost,sanitation,unimp_rural,2022,PFLO:HBLE,6004,70070,cb:wali:technical_cost:sanitation:unimp_rural:...
4,PFLO:HBLE,0,uganda,8,pop_unimproved_rural,1.774752e+07,1.774752e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0,cb,wali,technical_cost,sanitation,unimp_rural,2023,PFLO:HBLE,6004,70070,cb:wali:technical_cost:sanitation:unimp_rural:...


In [39]:
cb_data.strategy.unique()

array(['PFLO:HBLE', 'PFLO:NDC_25', 'NDC_25_UNCONDITIONAL',
       'PFLO:NDC_25_UNCONDITIONAL_CE'], dtype=object)

In [40]:
cb_data.strategy_code.unique()

array(['PFLO:HBLE', 'PFLO:NDC_25', 'PFLO:NDC_25_UNCONDITIONAL',
       'PFLO:NDC_25_UNCONDITIONAL_CE'], dtype=object)

In [41]:
# --- Read GDP and merge ---
# Keep only needed columns
gdp = ssp_data.loc[:, ["primary_id", "time_period", "gdp_mmm_usd"]]
gdp.head(100)

,primary_id,time_period,gdp_mmm_usd
0,0,4,96.640000
1,0,5,99.490000
2,0,6,103.010000
3,0,7,107.730000
4,0,8,113.483902
...,...,...,...
95,70070,47,885.672553
96,70070,48,903.386004
97,70070,49,921.453724
98,70070,50,939.882799


In [42]:
gdp[gdp['primary_id']== 77077]

,primary_id,time_period,gdp_mmm_usd
156,77077,4,96.640000
157,77077,5,99.490000
158,77077,6,103.010000
159,77077,7,107.730000
160,77077,8,113.483902
161,77077,9,120.406420
162,77077,10,128.353244
163,77077,11,137.787207
164,77077,12,150.256950
165,77077,13,165.012182


In [43]:
# Merge cb_data with gdp on primary_id and time_period
cb_data = cb_data.merge(gdp, on=["primary_id", "time_period"], how="left")
cb_data

,strategy_code,future_id,region,time_period,difference_variable,variable_value_baseline,variable_value_pathway,difference_value,variable,value,...,sector,cb_type,item_1,item_2,Year,strategy,strategy_id,primary_id,ids,gdp_mmm_usd
0,PFLO:HBLE,0,uganda,4,pop_unimproved_rural,1.932346e+07,1.932346e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.000000,...,wali,technical_cost,sanitation,unimp_rural,2019,PFLO:HBLE,6004,70070,cb:wali:technical_cost:sanitation:unimp_rural:...,96.640000
1,PFLO:HBLE,0,uganda,5,pop_unimproved_rural,1.951974e+07,1.951974e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.000000,...,wali,technical_cost,sanitation,unimp_rural,2020,PFLO:HBLE,6004,70070,cb:wali:technical_cost:sanitation:unimp_rural:...,99.490000
2,PFLO:HBLE,0,uganda,6,pop_unimproved_rural,1.896047e+07,1.896047e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.000000,...,wali,technical_cost,sanitation,unimp_rural,2021,PFLO:HBLE,6004,70070,cb:wali:technical_cost:sanitation:unimp_rural:...,103.010000
3,PFLO:HBLE,0,uganda,7,pop_unimproved_rural,1.837001e+07,1.837001e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.000000,...,wali,technical_cost,sanitation,unimp_rural,2022,PFLO:HBLE,6004,70070,cb:wali:technical_cost:sanitation:unimp_rural:...,107.730000
4,PFLO:HBLE,0,uganda,8,pop_unimproved_rural,1.774752e+07,1.774752e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.000000,...,wali,technical_cost,sanitation,unimp_rural,2023,PFLO:HBLE,6004,70070,cb:wali:technical_cost:sanitation:unimp_rural:...,113.483902
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
45203,PFLO:NDC_25_UNCONDITIONAL_CE,0,uganda,51,frac_scoe_heat_energy_residential_solid_biomass,NaN,NaN,NaN,cb:scoe:indoor_air_pollution,0.067919,...,scoe,indoor_air_pollution,None,None,2066,PFLO:NDC_25_UNCONDITIONAL_CE,6012,78078,cb:scoe:indoor_air_pollution:6012,958.680455
45204,PFLO:NDC_25_UNCONDITIONAL_CE,0,uganda,52,frac_scoe_heat_energy_residential_solid_biomass,NaN,NaN,NaN,cb:scoe:indoor_air_pollution,0.066635,...,scoe,indoor_air_pollution,None,None,2067,PFLO:NDC_25_UNCONDITIONAL_CE,6012,78078,cb:scoe:indoor_air_pollution:6012,977.854064
45205,PFLO:NDC_25_UNCONDITIONAL_CE,0,uganda,53,frac_scoe_heat_energy_residential_solid_biomass,NaN,NaN,NaN,cb:scoe:indoor_air_pollution,0.064921,...,scoe,indoor_air_pollution,None,None,2068,PFLO:NDC_25_UNCONDITIONAL_CE,6012,78078,cb:scoe:indoor_air_pollution:6012,997.411145
45206,PFLO:NDC_25_UNCONDITIONAL_CE,0,uganda,54,frac_scoe_heat_energy_residential_solid_biomass,NaN,NaN,NaN,cb:scoe:indoor_air_pollution,0.062852,...,scoe,indoor_air_pollution,None,None,2069,PFLO:NDC_25_UNCONDITIONAL_CE,6012,78078,cb:scoe:indoor_air_pollution:6012,1017.359368


In [44]:
96.640000/35.35

2.7338048090523337

In [45]:
cb_data[['time_period','gdp_mmm_usd']]

# 2019: 96.640000 #PPP
# 2019: 35.35 #USD
# 2.7
# https://data.worldbank.org/indicator/NY.GDP.MKTP.CD?locations=UG

,time_period,gdp_mmm_usd
0,4,96.640000
1,5,99.490000
2,6,103.010000
3,7,107.730000
4,8,113.483902
...,...,...
45203,51,958.680455
45204,52,977.854064
45205,53,997.411145
45206,54,1017.359368


In [46]:
cb_data.isna().sum()

strategy_code                 0
future_id                     0
region                        0
time_period                   0
difference_variable           0
variable_value_baseline     208
variable_value_pathway      364
difference_value            208
variable                      0
value                         0
name                          0
sector                        0
cb_type                       0
item_1                      208
item_2                     3340
Year                          0
strategy                      0
strategy_id                   0
primary_id                    0
ids                           0
gdp_mmm_usd                   0
dtype: int64

In [47]:
# Save the results
cb_data.to_csv(os.path.join(CB_OUTPUT, "cba_data_air_pollution.csv"), index = False)


In [48]:
cb_data

,strategy_code,future_id,region,time_period,difference_variable,variable_value_baseline,variable_value_pathway,difference_value,variable,value,...,sector,cb_type,item_1,item_2,Year,strategy,strategy_id,primary_id,ids,gdp_mmm_usd
0,PFLO:HBLE,0,uganda,4,pop_unimproved_rural,1.932346e+07,1.932346e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.000000,...,wali,technical_cost,sanitation,unimp_rural,2019,PFLO:HBLE,6004,70070,cb:wali:technical_cost:sanitation:unimp_rural:...,96.640000
1,PFLO:HBLE,0,uganda,5,pop_unimproved_rural,1.951974e+07,1.951974e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.000000,...,wali,technical_cost,sanitation,unimp_rural,2020,PFLO:HBLE,6004,70070,cb:wali:technical_cost:sanitation:unimp_rural:...,99.490000
2,PFLO:HBLE,0,uganda,6,pop_unimproved_rural,1.896047e+07,1.896047e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.000000,...,wali,technical_cost,sanitation,unimp_rural,2021,PFLO:HBLE,6004,70070,cb:wali:technical_cost:sanitation:unimp_rural:...,103.010000
3,PFLO:HBLE,0,uganda,7,pop_unimproved_rural,1.837001e+07,1.837001e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.000000,...,wali,technical_cost,sanitation,unimp_rural,2022,PFLO:HBLE,6004,70070,cb:wali:technical_cost:sanitation:unimp_rural:...,107.730000
4,PFLO:HBLE,0,uganda,8,pop_unimproved_rural,1.774752e+07,1.774752e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.000000,...,wali,technical_cost,sanitation,unimp_rural,2023,PFLO:HBLE,6004,70070,cb:wali:technical_cost:sanitation:unimp_rural:...,113.483902
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
45203,PFLO:NDC_25_UNCONDITIONAL_CE,0,uganda,51,frac_scoe_heat_energy_residential_solid_biomass,NaN,NaN,NaN,cb:scoe:indoor_air_pollution,0.067919,...,scoe,indoor_air_pollution,None,None,2066,PFLO:NDC_25_UNCONDITIONAL_CE,6012,78078,cb:scoe:indoor_air_pollution:6012,958.680455
45204,PFLO:NDC_25_UNCONDITIONAL_CE,0,uganda,52,frac_scoe_heat_energy_residential_solid_biomass,NaN,NaN,NaN,cb:scoe:indoor_air_pollution,0.066635,...,scoe,indoor_air_pollution,None,None,2067,PFLO:NDC_25_UNCONDITIONAL_CE,6012,78078,cb:scoe:indoor_air_pollution:6012,977.854064
45205,PFLO:NDC_25_UNCONDITIONAL_CE,0,uganda,53,frac_scoe_heat_energy_residential_solid_biomass,NaN,NaN,NaN,cb:scoe:indoor_air_pollution,0.064921,...,scoe,indoor_air_pollution,None,None,2068,PFLO:NDC_25_UNCONDITIONAL_CE,6012,78078,cb:scoe:indoor_air_pollution:6012,997.411145
45206,PFLO:NDC_25_UNCONDITIONAL_CE,0,uganda,54,frac_scoe_heat_energy_residential_solid_biomass,NaN,NaN,NaN,cb:scoe:indoor_air_pollution,0.062852,...,scoe,indoor_air_pollution,None,None,2069,PFLO:NDC_25_UNCONDITIONAL_CE,6012,78078,cb:scoe:indoor_air_pollution:6012,1017.359368
